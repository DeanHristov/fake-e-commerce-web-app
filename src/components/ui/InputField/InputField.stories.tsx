import type { Meta, StoryObj } from '@storybook/react';
import InputField, { IInputFieldProps } from './index';
import { MagnifyingGlassIcon } from '@heroicons/react/24/solid';

const meta: Meta<IInputFieldProps> = {
  title: 'UI/InputField',
  component: InputField,
  argTypes: {
    onChange: { action: 'onChange():' },
    type: {
      options: ['text', 'password'],
      control: { type: 'select' },
    },
  },
  parameters: {
    layout: 'centered',
  },
};

type Story = StoryObj<IInputFieldProps>;

export const Primary: Story = {
  args: {},
};

export const WithLeftIcon: Story = {
  args: {
    leftIcon: <MagnifyingGlassIcon className="w-7 h-7 text-zinc-100" />,
  },
};

export const WithRightIcon: Story = {
  args: {
    rightIcon: <MagnifyingGlassIcon className="w-7 h-7 text-zinc-100" />,
  },
};

export default meta;
