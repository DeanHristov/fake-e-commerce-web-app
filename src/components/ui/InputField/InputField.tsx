import { ChangeEvent, FC, ReactNode, useEffect, useState } from 'react';

export type TInputFieldTypes = 'text' | 'password';

export interface IInputFieldProps {
  onChange: (value: string) => void;
  type?: TInputFieldTypes;
  leftIcon?: ReactNode;
  rightIcon?: ReactNode;
  placeholder?: string;
}

const SearchField: FC<IInputFieldProps> = ({
  type = 'text',
  placeholder,
  onChange,
  leftIcon,
  rightIcon,
  ...rest
}) => {
  const [value, setValue] = useState<string>('');

  useEffect(() => onChange && onChange(value), [value]);

  return (
    <div className="relative w-full rounded-lg h-10 border border-gray-500">
      {leftIcon && (
        <div className="absolute top-1 left-2 right-auto">{leftIcon}</div>
      )}
      <input
        type={type}
        value={value}
        onChange={(ev: ChangeEvent<HTMLInputElement>) =>
          setValue(ev.target.value)
        }
        placeholder={placeholder}
        className={`h-full ${
          leftIcon ? 'pl-10 pr-6' : 'px-6'
        } w-full rounded-lg bg-gray-200 text-slate-700`}
        {...rest}
      />
      {rightIcon && (
        <div className="absolute top-1 left-auto right-2">{rightIcon}</div>
      )}
    </div>
  );
};

export default SearchField;
