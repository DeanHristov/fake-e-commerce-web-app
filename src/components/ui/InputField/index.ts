import InputField, { IInputFieldProps } from './InputField';

export type { IInputFieldProps };

export default InputField;
