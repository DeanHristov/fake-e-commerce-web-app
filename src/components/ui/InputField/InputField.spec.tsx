import { fireEvent, render, waitFor } from '@testing-library/react';
import InputField from './index';

describe('UI/Component <InputField {...} />', () => {
  it('Should render a default state of the Component', () => {
    const { container } = render(
      <InputField
        type="text"
        onChange={jest.fn()}
        placeholder={'Text me...'}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Should trigger onChange event during typing', async () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(
      <InputField
        type="text"
        onChange={mockOnChange}
        data-testid="input-field"
        placeholder={'Text me...'}
      />,
    );

    const $input = getByTestId('input-field');
    await waitFor(() =>
      fireEvent.change($input, { target: { value: 'demo-user@example.com' } }),
    );

    expect(mockOnChange).toBeCalledWith('demo-user@example.com');
  });
});
