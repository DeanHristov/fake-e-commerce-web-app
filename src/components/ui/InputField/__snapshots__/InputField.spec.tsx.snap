// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Component <InputField {...} /> Should render a default state of the Component 1`] = `
<div>
  <div
    class="relative w-full rounded-lg h-10 border border-gray-500"
  >
    <input
      class="h-full px-6 w-full rounded-lg bg-gray-200 text-slate-700"
      placeholder="Text me..."
      type="text"
      value=""
    />
  </div>
</div>
`;
